fea-rs/src/compile/tables/var_store.rs

```diff
@@ -0,0 +1,312 @@
+        let mut key_lookup = VariationStoreKeyLookup::default();
+        let mut region_list = self
+            .all_regions
+            .into_iter()
+            .map(|(reg, idx)| (idx, reg))
+            .collect::<Vec<_>>();
+        region_list.sort_unstable();
+        let region_list = region_list.into_iter().map(|(_, reg)| reg).collect();
+        let mut item_variations = Vec::new();
+
+        // Now: for each region set, we will produce one or more subtables
+        for (region, deltas) in self.delta_sets {
+            add_deltas(region, deltas, &mut item_variations, &mut key_lookup);
+        }
+
+        let varstore =
+            ItemVariationStore::new(VariationRegionList::new(region_list), item_variations);
+        (varstore, key_lookup)
+    }
+}
+
+/// create one or more subtables for the deltas in a particular region
+fn add_deltas(
+    regionset: RegionSet,
+    deltas: IndexMap<DeltaSet, DeltaKey>,
+    out: &mut Vec<Option<ItemVariationData>>,
+    keys: &mut VariationStoreKeyLookup,
+) {
+    // if there are no deltas, we push an empty subtable
+    if deltas.is_empty() {
+        out.push(None);
+        return;
+    }
+
+    let n_regions = regionset.indicies.len();
+
+    // for now we just always use long indices
+    let word_delta_count: u16 = n_regions as _;
+
+    // (item_count, BE bytes)
+    let mut finished = Vec::new();
+    let mut current = Vec::new();
+
+    // the index of the current subtable
+    let mut outer = out.len() as u16;
+    // the index of the deltaset, in that subtable
+    let mut inner = 0u16;
+    for (delta, key) in deltas {
+        if inner == 0 {
+            // next_index wraps, so if it's zero and this is non-empty it means
+            // we just filled a subtable
+            if !current.is_empty() {
+                finished.push((0xffff, std::mem::take(&mut current)));
+                outer += 1;
+            }
+        }
+        let real_key = DeltaKey { outer, inner };
+        current.extend(delta.values.iter().flat_map(|v| (*v as i16).to_be_bytes()));
+        keys.set(key, real_key);
+        inner = inner.wrapping_add(1);
+    }
+
+    // add all subtables, or null subtables if they were empty
+    // (empty is only possible when the number of deltasets is divisible by 0xffff)
+    for (item_count, deltas) in finished.into_iter().chain(Some((inner, current))) {
+        if deltas.is_empty() {
+            out.push(None);
+        } else {
+            out.push(Some(ItemVariationData::new(
+                item_count,
+                word_delta_count,
+                regionset.indicies.clone(),
+                deltas,
+            )));
+        }
+    }
+}
+
+impl DeltaSet {
+    fn push(&mut self, val: i32) {
+        self.values.push(val)
+    }
+}
+
+impl VariationStoreKeyLookup {
+    fn set(&mut self, from: DeltaKey, to: DeltaKey) {
+        self.map.insert(from, to);
+    }
+
+    #[cfg(test)]
```

Comment:
do I understand correctly this VariationStoreKeyLookup is only actually needed when testing but a client that wants to call VariationStoreBuilder::build() doesn't actually care about this (for it can't call `get` anyway unless `#[cfg(test)]` is set which is usually just the case when running cargo test.

In that case, should this _not_ be returned by build() method but instead you add another method that the tests can call after build to get this VariationStoreKeyLookup object from the builder?

Reply:
+1 to hide this if possible

Reply:
This type keeps track of the mapping from temporary->final `VariationIndex` values, which need to get updated after we generate the table; this is implemented in #187.